### DISCUSSION TITLE
Added: LDM Notes for May 27, 2020

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#3526] [General]
[C# Language Design Notes for May 27, 2020](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-05-27.md)

Record syntax
    1. Record structs?
    2. Record syntax/keyword
    3. Details on property shorthand syntax

**Comment by HaloFour** [comment]
> Do not support structs for now. They already support many features of records and we can add more, time permitting.

This is all fine or good, but deciding to make `record` it's own flavor of type instead of a modifier of existing types will only back the team into an awkward corner later if it is decided to extend some of the behaviors of `records` to structs.  It also completely avoids the mentioned ambiguity and breaking change, and even if the ambiguity could be avoided in the compiler (by switching on the language version?  _really?_).  It's one additional keyword, one that could be written by auto-complete, and it's worth it for the sake of clarity and future flexibility.

So, based on these decisions, what does the following mean?

```cs
public struct Point(int X, int Y);
```

**Comment by agocke** [comment]
> awkward corner later if it is decided to extend some of the behaviors of records to structs

It was mentioned that the syntax `record struct Point(int X, int Y)` could be added later. Some concern that this is asymmetric, but doesn't seem to be a dealbreaker.

> public struct Point(int X, int Y);

As of last LDM, this is invalid syntax. When we have a complete primary constructor (not records) proposal, that will probably be a parallel construction to a non-record class with a primary constructor.

**Comment by HaloFour** [comment]
@agocke 

> It was mentioned that the syntax `record struct Point(int X, int Y)` could be added later. Some concern that this is asymmetric, but doesn't seem to be a dealbreaker.

Yes, it could, but now `struct` is a _modifier_ instead of a type in it's own right?  Or `record` is sometimes a modifier and sometimes a type?  That just feels weird.  Might as well go the F# route and make it an attribute for that extra bolted-on feeling.  So many issues could be outright avoided by making `record` a modifier and leaving everything else alone.

Especially this notion that the following could mean two completely different things in C# 8.0 vs C# 9.0.  Seems very unlike how C# has ever evolved contextual keywords.  I don't think that the move to evolve the language with the runtime justifies it.

```cs
public class C {
    public record M(int x, int y) { }
}
```

**Comment by agocke** [comment]
> Yes, it could, but now struct is a modifier instead of a type in it's own right? 

You mean from the syntax perspective? That's not how I would implement it, I would just make it an optional token after the `record` keyword. 

> Especially this notion that the following could mean two completely different things in C# 8.0 vs C# 9.0.

Yup, as mentioned this is a breaking change, and will be interpreted one way before C# 9 and a different way after.

**Comment by qrli** [comment]
> ```cs
> public record Point(int X, int Y);
> public record Point
> {
>    data int X;
>    data int Y;
> }
> ```
I understand this makes nominal record more consistent with positional record. But suppose if I were to change Y to have explicit property get/init implementation, what changes should I do? Like below?
```cs
    // data int Y;
    int Y 
    {
        get => backingFieldOfY; 
        init => backingFieldOfY = value >= 0 ? value : throw new ArgumentException("..."); 
    }
```

Or, make the question in another way: does it mean `data int Y` is another way to say `int Y { get; init; }`?

**Comment by agocke** [comment]
@qrli  Yes, `data` is a short-hand for init-only auto-prop

**Comment by iam3yal** [comment]
I know that what I'm going to propose here might be a bit weird but wouldn't it be possible to introduce `record class` later when `record struct` is introduced for symmetry and maybe for people who might want to be more explicit? so `record` gonna be a syntactic sugar to `record class`.

Personally, I prefer `record class` over `record` for consistency and because of that I wonder whether it might be sensible to add it later as suggested above.

**Comment by agocke** [comment]
I see no reason why we couldn't choose to do that.

**Comment by BreyerW** [comment]
I second all @HaloFour 's concerns. Since he already spilled everything i thought of i will leave it at that

**Comment by Unknown6656** [comment]
I agree with @eyalsk:
Introduce `record [class|struct] ....` instead of introducing a breaking change.

-------

![image](https://user-images.githubusercontent.com/8807985/83400387-05f4ae00-a403-11ea-9394-eb81a284e979.png)
_C# LDT meeting [ca. 2020, colorized]_

<sub>'twas just a joke 😘</sub>

**Comment by spydacarnage** [comment]
Joking aside, I totally agree.  Not having a breaking change, and having
the ability to potentially avoid somebody suggesting "valuerecord" or
some-such in the future, is well worth having to write "record class" vs
"record".

Also, while structs currently have value equality, the ability to easily
add "with" support to a struct would be something I'd like to see.

**Comment by Joe4evr** [comment]
> the ability to easily add "with" support to a struct would be something I'd like to see.

That would be as easy as this, once the work for the factory method attribute gets done:
```cs
public struct S
{
    [Factory]
    public S With() => this;
}
```
Though this does presume that any fields of reference types can't be mutated in-place. If there is such a case, you'd do `new S(this)` that makes a deeper copy.

**Comment by 333fred** [comment]
We don't even need copy constructors to implement with on structs. We can just dup them and then set fields/properties.

**Comment by MgSam** [comment]
I don't understand the LDT's seeming obsession with shaving off every possible character for the records feature. Is `record class` vs `record` really such a burden? Is `public int Age {get; init;}` vs `data int Age;` really such a burden? 

Snippets/autocomplete can handle most of the typing, and adding/changing types/members is not something most people are doing on a daily basis. You spend most of your time working with the types, not writing their definitions.

This isn't like adding `var` to the language where it's saving a ton of typing. The few saved characters are not worth the increased complexity and ambiguity to the language. Making records look like a 3rd fundamental kind of type is a mistake that will confuse newcomers and oldcomers alike. `data` saves so few characters it is not worth the ambiguity that it looks a lot like a field.

**Comment by HaloFour** [comment]
I'm conflicted about the notion of the `data` keyword and nominal records.  I can sorta understand the desire to make the declarations shorter, especially compared to their positional brethren.  But at the same time I think it obscures what the record is about and it limits what you can do now and potentially in the future with those members.

IMO, I'm not offended by having to have full and proper property declarations in a nominal record.  They are, afterall, the focal point of what that type represents.  Having them look and act like regular auto-properties would be fine to me and I think would decrease the cognitive load when learning about these types.  What I don't want to have to write is all of the rest of the boilerplate code that goes into value equality and cloning.  That's where I personally find the value of records.  Not in trying to make the data declaration as terse as humanly possible.  The `data` keyword kind of feels like an example of [trying to make a new language feature stand out](http://gafter.blogspot.com/2017/06/making-new-language-features-stand-out.html).

**Comment by MgSam** [comment]
I think that post from Neal makes a great point. 

I hate to always be the one beating this dead horse but `data` reminds me of the `in` keyword again. Shaves off a few characters at the expense of causing widespread confusion. 

I feel like the old maxim whereby every feature starts at -1000 points is not really followed on the LDT anymore. Adding new keywords, contextual or not, should definitely start at -1000 points. I can't see any scenario where `data` saving a few characters is worth 1001 points.

**Comment by CyrusNajmabadi** [comment]
> I don't understand the LDT's seeming obsession

Please don't editorialize.  We received a lot of feedback on this topic.  So we talked about it and considered it and even vote don our preference here from all the members.  The pros/cons were discussed in depth and were understood.  Individuals may not have the same weighting of goals and desires as you.  But that's that.  Language design is enormously subjective, and that means you may not end up liking how that lands on particular aspects compared to your own personal preferences.

**Comment by CyrusNajmabadi** [comment]
>  I can't see any scenario where data saving a few characters is worth 1001 points.

There is nothign then that could be said to change your mind here.  The LDM evaluated this and felt differently.  C'est la vie :)

**Comment by CyrusNajmabadi** [comment]
```c#
public class C {
    public record M(int x, int y) { }
}
```

this would not have been legal code before.  so it's a case of illegal code becoming legal.  There are *narrower* cases of true ambiguity, but we discussed and considered them, and felt they were unlikely enough, and the value here was high enough to warrant the opt-in change from teh user.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> but we discussed and considered them, and felt they were unlikely enough, and the value here was high enough to warrant the opt-in change from teh user.

To be absolutely clear, the "value" here is being able to omit the keyword `class`.  Six characters (with whitespace).  Likely 1-2 keystrokes via auto-complete.  Nothing more.

**Comment by YairHalberstadt** [comment]
I feel like the compromise of making `data` turn a field into an autoprop is the worst of all worlds:

- still looks like a field, not a prop.
- simplest thing to write is a field, so does not encourage properties by default.
- not much shorter than an autoprop
- introduces new contextual keyword
- still makes declarations in records completely different from other classes since `data` is only meaningful in a record.

**Comment by CyrusNajmabadi** [comment]
> To be absolutely clear, the "value" here is being able to omit the keyword class. Six characters (with whitespace). Likely 1-2 keystrokes via auto-complete. Nothing more.

I disagree.  The value is being able to think of record as an entity on it's own, not a modified "class".  i.e. i don't think of a "tuple" as a "tupleized struct" i think of it as a tuple.  I don't think of "nullable" as a "struct Nullable".  i just think of it as a nullable.  In this case, the balance leaned more toward people prefering 'record' as it's on entity concept.

> still makes declarations in records completely different from other classes since data is only meaningful in a record.

I don't believe that's true.  My recollection was that thsi was valid for properties in any context.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> i.e. i don't think of a "tuple" as a "tupleized struct" i think of it as a tuple. I don't think of "nullable" as a "struct Nullable". i just think of it as a nullable.

You don't declare either as a type so I don't see how the analogy is relevant.

> In this case, the balance leaned more toward people prefering 'record' as it's on entity concept.

Until you add back `record struct` in which case it becomes some weird hybrid and doesn't fit well in either model.

A "record" is still a class.  That class aspect is still the most important aspect about it.  It's still nullable.  It's still copy-by-reference.  A record _is_ a modified class, and probably eventually also a modified struct.

**Comment by CyrusNajmabadi** [comment]
> You don't declare either as a type so I don't see how the analogy is relevant.

Because i'm pointing out htat i don't think of this as  'recordized class'.  So it's not about omitting the 'class' for hte sake of omitting it.  It's taht i don't think if it as teh same thing.  The same way i don't think about a tuple as being the same as other types, or nullables, etc.  Yes, at the end of the day, they compile down to classes/structs.  But that doesn't mean i think that when i declare the type i should have to say that. 

You disagree.  That's fine.  I'm just pointing out that your assessment of what the 'savings' here is, or what guided teh decision for some is not accurate.

> A "record" is still a class.

A record compiles down to a class.  I don't think of it as a class.  This was something we talked about *this very day* in the ldm.   The different perspective are valid, but that doesn't mean that it's invalid to want to think of 'record' as an entity in its own right.

**Comment by CyrusNajmabadi** [comment]
> Until you add back record struct in which case it becomes some weird hybrid and doesn't fit well in either model.

As i stated in gitter a few times, my position is that i simply *want* structs to be records.  i.e. i wouldn't expect `record struct`.  I would expect to just enhance `struct` so they get all the goodness automatically that we're having to bring to `record` for reference types.

in other words, we already have ValueType values, taht are very close to what i want values to be.  I think we can get there with very little changes (and without having to say `record`).  I would like the same for `reference type` values, and that's what `record` is.  it's basically saying: have a reference type, but make it behave like all the ways values should behave, without me having to write all that in.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Which backs the design further into the corner.  If every struct has to be a record, then records can never be more than the opinions that can be applied to a naked struct.  Case in point, primary constructors.  If a struct is a record, then a primary constructor has to take on all of the opinions of auto-generating public members and deconstruction, regardless of whether or not that makes sense in that scenario.  It forces that non-record primary constructors can't be added to the language because there is no mechanism through which you can apply them to structs, because structs can never again be non-records.

It also implies very few changes can be made to the semantics of existing struct syntax to align them with records without subtly breaking tons of existing code.  And it implies that the difference won't impact all of the other features still expected to be designed on top of records, including discriminated unions.

Anyway, we clearly won't agree.  My money is on the first question being, "where's struct records?" and the answer, "structs are basically already records," not being well received.

**Comment by qrli** [comment]
@CyrusNajmabadi From June 1st LDM notes:
> A: For the first record in a type hierarchy, we should define equality based on all the accessible fields, including inherited ones, in the record.

So does it mean that a record may derive from a non-record class? If record is to be treated as a separate concept than class, as you said, it should not be allowed to derive from class IMO.

PS: And on the `data` keyword, may I ask your view point on why it is needed?

**Comment by CyrusNajmabadi** [comment]
> may I ask your view point on why it is needed?

Sure. It feels like something super common and we'd like that to be brief. `int Food;`is as short as currently possible, but that's already a private field. So we wanted something very minor in top of that too get a public/initable property. Hence `data` was borne.

**Comment by CyrusNajmabadi** [comment]
>  My money is on the first question being, "where's struct records?" and the answer, "structs are basically already records," not being well received.

Why?  We would be able to say: you don't need to do anything here. Want to use `with`?  You already can for all your structs because they already get this goodness. Only ref-types need it because it's such a far (and incompatible) distance with what you get with classes today.

**Comment by YairHalberstadt** [comment]
If the ldm feels that auto props are too long, and that `{ get; init; }` is a good default for a prop, how about using `int Prop {}` as a shorthand for `int Prop { get; init; }`.

I would rather just have it in full, but I would much rather `int Prop {}` over `data int Prop`. At least it looks like a property.

**Comment by CyrusNajmabadi** [comment]
> At least it looks like a property.  

Sure.  seems like something we could consider.

**Comment by ghord** [comment]
So just to be clear: is there a difference at all when using `data int Prop;` and `public int Prop { get; init; }`? Does it influence code generation for records feature at all?

If not, I  think it is not worth it to introduce this keyword just to avoid property syntax which c# programmers are very used to.

**Comment by SingleAccretion** [comment]
> `int Prop {}` as a shorthand for `int Prop { get; init; }`

On that note, It seems to me that replacing `data` with `init` might be worth considering (for the record though, I agree with @YairHalberstadt that `data` does feel like a sad compromise). So, we would write `init int Length;` instead of `data int Length;`. In my mind it has two advantages:
- One less keyword -> one less thing to remember.
- More intuitive what it does since it desugars into something with `init` in it.

It does still look like a field, but that seems unavoidable with this design (you could of course think of `int Length { data; }` being the same as `int Length { get; init; }` but that does not seem worth it).

Also, for what it's worth (not much, I know), I made these measurements:
| Keystrokes (`KeyDown` + `KeyUp` and `KeyDown` for modifiers) | Raw | With autocomplete |
|--------------------------|------|----------------------|
| `data int Length;`          | 18    | 15                           |
| `int Length { get; init; }` | 31   | 27                           |

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> Why? We would be able to say: you don't need to do anything here. Want to use with? You already can for all your structs because they already get this goodness. Only ref-types need it because it's such a far (and incompatible) distance with what you get with classes today.

Because value identity is only a part of that equation.  Even F# records add implementations to some half-dozen interfaces.  I'd expect the same in C#.  Not to mention equality operators.  Is this why adding them has been "controversial"?  And then there's the previously mentioned primary constructors.  I wouldn't expect (nor want) primary constructors on non-records to expose any members or deconstruction.

**Comment by iam3yal** [comment]
I prefer to type properties in full. `data type MyProp;` tells me nothing much like `type MyProp {}` I mean I don't know why would either of them _deserve_ to translate into `{get; init;}` it feels so arbitrary, imo something like the following is much better `init type MyProp;`, `init type MyProp {};` or `type MyProp {init;}` over the previous suggestions, depends on the LDT goals.

I feel like auto properties are short enough that it isn't really an issue so why even think about it? we have intellisense, snippets, refactoring tools for the lazy of us so introducing a new keyword and make it even shorter is a slippery slope that can be avoided.

**Comment by CyrusNajmabadi** [comment]
>  tells me nothing much like `type MyProp {}`

I think the problem there is that there's nothing to indicate accessibility.  i.e. if you take a property, and remove things, it doesn't feel like it should convey more.  `data` on hte other hand, is additive, and can mean it's adding several things.

**Comment by YairHalberstadt** [comment]
It feels very odd for a modifier to affect accessibility. It would be far less odd for everything in a record to be public by default. That's already the case for interfaces, and if you want to view `record` as a kind in and of itself, not a modifier on `class`, you should embrace it, and not assume defaults will be the same as classes rather than interfaces.

**Comment by agocke** [comment]
> It feels very odd for a modifier to affect accessibility

You mean like `public`, `private`, `protected`, and `internal`? 😉

**Comment by YairHalberstadt** [comment]
You know what I mean :-)

**Comment by iam3yal** [comment]
> `data` on hte other hand, is additive, and can mean it's **adding several things**.

Yes but it feels wrong when it adds several **unrelated** things and shoehorn them into an arbitrary  keyword, I mean does `data` imply `public`? does it imply a property? and if so does it imply a getter and an initializer? now you can always say yes that's our definition for it but it seems too abstract and detached from the idea it aims to represent so why waste such a nice keyword over something that is already trivial? if part of the concern is repeating the access modifier then like @YairHalberstadt said just make everything public by default which is probably the common use-case anyway.

**Comment by CyrusNajmabadi** [comment]
> I mean does data imply public? does it imply a property? and if so does it imply a getter and an initializer?

Sure :)

**Comment by HaloFour** [comment]
I don't recall any proposals seeking to make auto-properties less verbose.

**Comment by iam3yal** [comment]
[![Let It Be](https://i1.ytimg.com/vi/QDYfEBY9NM4/hqdefault.jpg)](https://www.youtube.com/watch?v=QDYfEBY9NM4 "Let It Be")

**Comment by MrJul** [comment]
> A record compiles down to a class. I don't think of it as a class. This was something we talked about _this very day_ in the ldm. The different perspective are valid, but that doesn't mean that it's invalid to want to think of 'record' as an entity in its own right.

The problem I see here is that *record as an entity* only applies to the declaration. Once a consumer is using the defined type, the difference between a class and a record disappears. If they are really different concepts, then new problems appear:

- I shouldn't be able to pass a record to a `T where T : class`. I also expect a new `where T : record` constraint to be added. Most existing methods thus won't work well with records. Do library authors have to add overloads supporting records everywhere they're supporting classes?
- What about other CLR languages, do they have to know that records shouldn't be treated as class and vice versa?
- If they're different entities, a class can't be derived from a record and a record can't be derived from a class.
- Existing record-like class implementations can't be simplified to records.
- When designing a new record, you have to be very careful, you won't be able to turn it to a fully fledged class if needed in the future without breaking all your consumers.

As I don't expect much of the above to be true, the *record as its own entity type* will likely be a very leaky abstraction. I believe it's different from tuple types where you can most of the time forget that they're compiled down to a `ValueTuple`. With records, the fact that it's compiled down to a class is not a detail. Unless all restrictions above are applied, which would probably make the feature unusable with the existing ecosystem.

Records started as a nice syntactic sugar to avoid writing a bunch of very common code, only impacting the declaration site. Imho, it shouldn't affect consumers at all. As a consumer of that nice `Point` type I found in a library, I shouldn't have to care whether it was implemented in C# 1 as a class or in C# 9 as a record, as long as it's documented to behave the same.

**Comment by Vlad-Stryapko** [comment]
I absolutely agree with the points above. I don't think the idea 'record is a separate entity' will fundamentally work out nicely; most likely, it will result in a leaky abstraction and people will have to remember that records are indeed classes under the hood. 
I wish I could think of a record as something on its own and I get where @CyrusNajmabadi is coming from, but I simply don't see it happening. For all intents and purposes a record seems to be just a class without a class keyword.

**Comment by bbarry** [comment]
Would all records share a base class like structs do (a base class that is not `object` itself)?

**Comment by HaloFour** [comment]
@bbarry 

Sounds like it based on the comments to the PR updating the records spec:

https://github.com/dotnet/csharplang/pull/3527#discussion_r434063198

IMO inheritance certainly complicates records, especially if a record can inherit from a non-record class.  It seems inappropriate for the derived class to not call the base `Equals`/`GetHashCode` members, but for `Object` those members define the identity of the type based on the reference.  So any derived record could never have value identity.

A compromise to this could be that if the base class (or any of its base classes) overrides `Equals`/`GetHashCode` that the compiler emits a call to those methods.  But if the call would be to `Object.Equals` or `Object.GetHashCode` that the compiler would not emit that call and it would not take the base identity into consideration.  This might sound a little weird on the surface, but when you call a `base` method in an overridden member the compiler has to resolve the most-specific base override for that method and call it directly, not virtually, so the compiler could certainly do that selectively.

**Comment by bbarry** [comment]
I guess I am just expecting some verbiage in the records proposal similar to that in the struct spec:

https://github.com/dotnet/csharplang/blob/53f957b472ee458a43844680f8f555694c4cb79a/spec/structs.md#struct-members

> ## Struct members
> 
> The members of a struct consist of the members introduced by its *struct_member_declaration*s and the members inherited from the type `System.ValueType`.
> 
> ```antlr
> struct_member_declaration
>     : constant_declaration
>     | field_declaration
>     | method_declaration
>     | property_declaration
>     | event_declaration
>     | indexer_declaration
>     | operator_declaration
>     | constructor_declaration
>     | static_constructor_declaration
>     | type_declaration
>     | struct_member_declaration_unsafe
>     ;
> ```
> 
> Except for the differences noted in [Class and struct differences](structs.md#class-and-struct-differences), the descriptions of class members provided in [Class members](classes.md#class-members) through [Iterators](classes.md#iterators) apply to struct members as well.

**Comment by YairHalberstadt** [comment]
@HaloFour 
That would suffer from the brittle base class problem.
If the intervening class later defines an equality operator, it will not be called.

**Comment by HaloFour** [comment]
@YairHalberstadt 

> That would suffer from the brittle base class problem.
> If the intervening class later defines an equality operator, it will not be called.

That's already the case.

**Comment by YairHalberstadt** [comment]
@HaloFour what do you mean?

**Comment by HaloFour** [comment]
@YairHalberstadt 

If you have the following:

```cs
public class A {
    public virtual void M() { }
}

public class B : A { }

public class C : B {
    public override void M() {
        base.M();
    }
}
```

The `base.M()` call calls `A.M` directly, as there is no method `B.M` to call.  If `B` lived in another assembly and someone added an override to `M` then `C.M` would not call it until it was recompiled.

If the runtime or language intends to change this then yes, perhaps the selective base call approach wouldn't be appropriate.  But this has been the behavior of C# and the CLR for 20 years.

**Comment by YairHalberstadt** [comment]
You're right:

https://docs.microsoft.com/en-us/archive/blogs/ericlippert/putting-a-base-in-the-middle

**Comment by HaloFour** [comment]
This seems to be what Scala is doing:

```scala
class Base1 {
  override def equals(that: Any): Boolean = false
}

class Base2 { }

case class Foo(val x: Int) extends Base1 { }
case class Bar(val x: Int) extends Base2 { }


object Main {
  def main(args: Array[String]): Unit = {
    
    if (Foo(123) == Foo(123)) {
      println("foo equals")
    } else {
      println("foo not equals")
    }
    
    if (Bar(123) == Bar(123)) {
      println("bar equals")
    } else {
      println("bar not equals")
    }
  }
}
```

prints:

```
foo not equals
bar equals
```

**Comment by CyrusNajmabadi** [comment]
>I shouldn't be able to pass a record to a T where T : class 

Sure you would. That's the constraint that says this is a reference type. We picked `class` there because it's convenient not because you're only restricted to classes. 

For example you can pass interfaces to your type parameter here, as well as delegates. Interface are not classes. Delegates aren't either. But they are things that satisfy the`class` constraint. 

 The same would hold for records.

**Comment by CyrusNajmabadi** [comment]
Similarly, the `struct` constraint doesn't mean structs or value types. That's why there are some structs you can't pass to methods that have that constraint. `class/struct` in constraints we're just really convenient wats for us to have a syntactic shorthand for the constraint concept here. It's never been a straight literally interpretation.

**Comment by CyrusNajmabadi** [comment]
> Do library authors have to add overloads supporting records everywhere they're supporting classes?

No. As per above. If they have overloads to handle reference types specially, then records will just work there.

**Comment by CyrusNajmabadi** [comment]
> Existing record-like class implementations can't be simplified to records.

Why not?

**Comment by CyrusNajmabadi** [comment]
> When designing a new record, you have to be very careful, you won't be able to turn it to a fully fledged class if needed in the future without breaking all your consumers.

Why not?  Our designs up to now have considered that an important migration path for code. I know of nothing in the design that would prevent or limit that. Indeed, I was planning on writing the ide features to do just this.

**Comment by omariom** [comment]
```C#
public record Point
{
   data int X;
   data int Y;
}
```
I would exchange `data` shortcut for struct records without thinking for a second.

**Comment by MrJul** [comment]
> Sure you would. That's the constraint that says this is a reference type. We picked `class` there because it's convenient not because you're only restricted to classes.

That's fair, a record is just another reference type and `where T : class` has always allowed any reference type despite its name, I'm fine with that, that ship has sailed.

> Why not? Our designs up to now have considered that an important migration path for code. I know of nothing in the design that would prevent or limit that. Indeed, I was planning on writing the ide features to do just this.

Then records are just classes, that's just not a compiler detail. For all intents and purposes records and classes are the same and interchangeable (and they really are under the hood). That's why others and I have difficulties agreeing to the concept of *records as their own kind of entity* and the drop of the `class` keyword (with a breaking change!) when even the tooling will help you convert from one to another.

When introducing tuples to people, they don't have to know how it's compiled: as you said I also think of a tuple as a separate entity and the implementation is a detail. With records, people will be taught that records are classes with sugar (a sweet one that is) since they *have* to know that to use them properly.

I really appreciate that you took the time to answer.

**Comment by CyrusNajmabadi** [comment]
> Then records are just classes,

As I said myself:

> "A record compiles down to a class."

Just as many other things in the language compile down to other things. 

> That's why others and I have difficulties agreeing to the concept of records as their own kind of entity 

That's a perfectly fine opinion. I don't share it. Something can be it's own entity, but still be compatible with another. Tuples are their own entity for me, but they compile down to a standard struct and can be used anywhere that specifies a value type constraint. That doesn't bother me.

**Comment by CyrusNajmabadi** [comment]
> I really appreciate that you took the time to answer.

Happy to :-)

**Comment by CyrusNajmabadi** [comment]
> When introducing tuples to people, they don't have to know how it's compiled:

But a lot of people did care about this. It was actually a large point if contention for people (and still is), that it's a mutable value type. 

You'll always have some people who care and pay attention to the underlying representation, and some who don't care.

**Comment by CyrusNajmabadi** [comment]
> With records, people will be taught that records are classes with sugar (a sweet one that is) since they have to know that to use them properly.

I have no problem with that. TypeScript classes (before ecma introduced them) compiled down into JavaScript constructor functions. It was just sweet sugar. So what? We could have gone with a `constructor` keyword on functions (which was considered), but we decided we preferred not to and that a dedicated (i.e. non-modified) syntax was preferable. 

If people get told that this is a nice simple shorthand for a class with lots of behavior baked in, I have no problem with that. And that doesn't make me not want a simple syntax here for these.

**Comment by AlgorithmsAreCool** [comment]
I am a little bummed to see struct support getting pushed out for now.

One important side effect of making type declarations easier/smaller, is that people will want to define more types now that it isn't as much of a pain.

Finer grained types can help the compiler assist in validating correct programs. But if they are all classes it holds this feature back from performance sensitive areas where chatty object allocation would impede performance. Which is doubly unfortunate, since performance oriented code often needs additional help in being proven correct.

My current workaround is to define my fine-grained types as struct records in F# then reference the assembly from my perfy C# code to try to have my cake and eat it too.

**Comment by ufcpp** [comment]
I wish primary constructor for (non-record) classes and structs will be in C# 9 and it is consistent with that for records.

**Comment by Unknown6656** [comment]
@ufcpp : I really hope that, too!
